File: src/scraper.js
import fetch from 'node-fetch';
import * as cheerio from 'cheerio';
import fs from 'fs';
import path from 'path';

const AAVSO_URL = 'https://apps.aavso.org/webobs/results/?star=t+crb&num_results=20&obs_types=all&page=1';

/**
 * Extrai os dados da observação mais recente da estrela T CrB
 * @returns {Promise<Object>} Objeto com os dados da observação
 */
async function scrapeLatestObservation() {
  try {
    console.log('Buscando dados da AAVSO...');
    const response = await fetch(AAVSO_URL);
    const html = await response.text();
    
    const $ = cheerio.load(html);
    
    // Encontra a primeira linha da tabela de observações (mais recente)
    const firstRow = $('table.observations tbody tr.obs').first();
    
    if (firstRow.length === 0) {
      throw new Error('Nenhuma observação encontrada na página');
    }
    
    // Extrai os dados das células
    // A tabela tem 3 colunas vazias no início (colspan="3")
    const cells = firstRow.find('td');
    
    const data = {
      star: $(cells[1]).text().trim(),           // Star (índice 1, após as 3 colunas vazias)
      jd: $(cells[2]).text().trim(),             // JD
      calendarDate: $(cells[3]).text().trim(),   // Calendar Date
      magnitude: $(cells[4]).find('a').text().trim() || $(cells[4]).text().trim(), // Magnitude
      error: $(cells[5]).text().trim(),          // Error
      filter: $(cells[6]).text().trim(),         // Filter
      observer: $(cells[7]).text().trim(),       // Observer
      timestamp: new Date().toISOString()
    };
    
    console.log('Dados extraídos:', data);
    return data;
    
  } catch (error) {
    console.error('Erro ao fazer scraping:', error.message);
    throw error;
  }
}

/**
 * Salva os dados em um arquivo CSV
 * @param {Object} data - Dados da observação
 * @param {string} filename - Nome do arquivo CSV
 */
function saveToCSV(data, filename = 'observations.csv') {
  const dataDir = path.join(process.cwd(), 'data');
  
  // Cria o diretório data se não existir
  if (!fs.existsSync(dataDir)) {
    fs.mkdirSync(dataDir, { recursive: true });
  }
  
  const filepath = path.join(dataDir, filename);
  const fileExists = fs.existsSync(filepath);
  
  // Cabeçalho do CSV
  const header = 'Star,JD,Calendar Date,Magnitude,Error,Filter,Observer,Timestamp\n';
  
  // Linha de dados
  const row = `"${data.star}","${data.jd}","${data.calendarDate}","${data.magnitude}","${data.error}","${data.filter}","${data.observer}","${data.timestamp}"\n`;
  
  if (!fileExists) {
    // Se o arquivo não existe, cria com cabeçalho
    fs.writeFileSync(filepath, header + row);
    console.log(`Arquivo criado: ${filepath}`);
  } else {
    // Se existe, apenas adiciona a nova linha
    fs.appendFileSync(filepath, row);
    console.log(`Dados adicionados ao arquivo: ${filepath}`);
  }
}

/**
 * Verifica se a observação já existe no CSV (evita duplicatas)
 * @param {Object} data - Dados da observação
 * @param {string} filename - Nome do arquivo CSV
 * @returns {boolean} True se já existe
 */
function observationExists(data, filename = 'observations.csv') {
  const filepath = path.join(process.cwd(), 'data', filename);
  
  if (!fs.existsSync(filepath)) {
    return false;
  }
  
  const content = fs.readFileSync(filepath, 'utf-8');
  // Verifica se já existe uma linha com o mesmo JD (Julian Date)
  return content.includes(`"${data.jd}"`);
}

/**
 * Função principal
 */
async function main() {
  try {
    const data = await scrapeLatestObservation();
    
    if (observationExists(data)) {
      console.log('Observação já existe no arquivo. Nenhuma ação necessária.');
    } else {
      saveToCSV(data);
      console.log('Nova observação salva com sucesso!');
    }
    
  } catch (error) {
    console.error('Erro na execução:', error);
    process.exit(1);
  }
}

// Executa se for chamado diretamente
if (import.meta.url === `file://${process.argv[1]}`) {
  main();
}

export { scrapeLatestObservation, saveToCSV, observationExists };

File: scraper.py
"""
Módulo de web scraping para coletar dados da estrela T CrB do portal AAVSO
"""

import requests
from bs4 import BeautifulSoup
from datetime import datetime
from typing import Dict, Optional
import re


def parse_calendar_date(calendar_date: str) -> str:
    """
    Converte data do formato AAVSO para dd/MM/YYYY HH:mm
    Exemplo: "2025 Oct. 26.07083" -> "26/10/2025 01:42"
    """
    try:
        # Parse: "2025 Oct. 26.07083"
        parts = calendar_date.split()
        year = parts[0]
        month_str = parts[1].replace('.', '')
        day_decimal = parts[2]
        
        # Converte mês para número
        months = {
            'Jan': '01', 'Feb': '02', 'Mar': '03', 'Apr': '04',
            'May': '05', 'Jun': '06', 'Jul': '07', 'Aug': '08',
            'Sep': '09', 'Oct': '10', 'Nov': '11', 'Dec': '12'
        }
        month = months.get(month_str, '01')
        
        # Separa dia e fração decimal
        day_parts = day_decimal.split('.')
        day = day_parts[0].zfill(2)
        
        # Converte fração decimal para horas e minutos
        if len(day_parts) > 1:
            fraction = float('0.' + day_parts[1])
            hours = int(fraction * 24)
            minutes = int((fraction * 24 - hours) * 60)
        else:
            hours = 0
            minutes = 0
        
        return f"{day}/{month}/{year} {hours:02d}:{minutes:02d}"
    except:
        # Se falhar, retorna a data original
        return calendar_date


class TCrbScraper:
    """Scraper para dados de observação da estrela T CrB"""
    
    AAVSO_URL = "https://apps.aavso.org/webobs/results/?star=t+crb&num_results=20&obs_types=all&page=1"
    
    def __init__(self):
        self.session = requests.Session()
        self.session.headers.update({
            'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36'
        })
    
    def fetch_latest_observation(self) -> Optional[Dict]:
        """
        Busca a observação mais recente da estrela T CrB
        
        Returns:
            Dict com os dados da observação ou None se houver erro
        """
        try:
            print(f"Buscando dados de {self.AAVSO_URL}...")
            response = self.session.get(self.AAVSO_URL, timeout=30)
            response.raise_for_status()
            
            soup = BeautifulSoup(response.text, 'html.parser')
            
            # Encontra a tabela de observações
            table = soup.find('table', class_='observations')
            if not table:
                raise ValueError("Tabela de observações não encontrada")
            
            # Encontra o tbody
            tbody = table.find('tbody')
            if not tbody:
                raise ValueError("Tbody não encontrado")
            
            # Encontra a primeira linha com classe 'obs'
            first_row = tbody.find('tr', class_='obs')
            if not first_row:
                raise ValueError("Nenhuma observação encontrada")
            
            # Extrai as células
            cells = first_row.find_all('td')
            
            # A tabela tem 3 colunas vazias no início (colspan="3")
            # Então os dados começam no índice 1
            # Extrai calendar_date original
            calendar_date_original = cells[3].get_text(strip=True)
            
            # Converte para formato brasileiro
            calendar_date_formatted = parse_calendar_date(calendar_date_original)
            
            # Formata a data de coleta no formato brasileiro
            scraped_at = datetime.now().strftime('%d/%m/%Y %H:%M')
            
            data = {
                'star': cells[1].get_text(strip=True),
                'jd': cells[2].get_text(strip=True),
                'calendar_date': calendar_date_formatted,  # Usa a data formatada
                'magnitude': cells[4].get_text(strip=True),
                'error': cells[5].get_text(strip=True),
                'filter': cells[6].get_text(strip=True),
                'observer': cells[7].get_text(strip=True),
                'scraped_at': scraped_at
            }
            
            print(f"✓ Dados extraídos: {data['star']} - Magnitude: {data['magnitude']}")
            return data
            
        except requests.RequestException as e:
            print(f"✗ Erro na requisição HTTP: {e}")
            return None
        except (ValueError, IndexError, AttributeError) as e:
            print(f"✗ Erro ao parsear HTML: {e}")
            return None
        except Exception as e:
            print(f"✗ Erro inesperado: {e}")
            return None


def main():
    """Função principal para teste"""
    scraper = TCrbScraper()
    data = scraper.fetch_latest_observation()
    
    if data:
        print("\n=== Dados Coletados ===")
        for key, value in data.items():
            print(f"{key:15}: {value}")
    else:
        print("\n✗ Falha ao coletar dados")
        return 1
    
    return 0


if __name__ == "__main__":
    exit(main())